### PR TITLE
[NN] Label Propagation & Directional Graph Networks

### PR DESCRIPTION
**Pull request by rudongyu**
## Description
Add the LabelPropagation module and DGNConv module.

Label propagation is a simple yet effective way for semi-supervised node classification, utilizing only the label information in graphs.

Directional Graph Networks allow a better representation of the anisotropic features in different physical or biological problems.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by mufeili** [comment]
Lint check failed

```
************* Module dgl.nn.pytorch.utils

python/dgl/nn/pytorch/utils.py:500:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/utils.py:509:0: C0304: Final newline missing (missing-final-newline)

************* Module dgl.nn.pytorch.conv.dgnconv

python/dgl/nn/pytorch/conv/dgnconv.py:87:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/dgnconv.py:90:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/dgnconv.py:93:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/dgnconv.py:125:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/dgnconv.py:128:0: C0303: Trailing whitespace (trailing-whitespace)

python/dgl/nn/pytorch/conv/dgnconv.py:89:0: C0301: Line too long (101/100) (line-too-long)

python/dgl/nn/pytorch/conv/dgnconv.py:173:4: W0231: __init__ method from base class 'PNAConv' is not called (super-init-not-called)

python/dgl/nn/pytorch/conv/dgnconv.py:5:0: C0411: standard import "from functools import partial" should be placed before "import torch" (wrong-import-order)
```

File: examples/pytorch/label_propagation/model.py [hunk 1]
```diff
@@ -41,16 +41,15 @@ def forward(self, g, labels, mask=None, post_step=lambda y: y.clamp_(0., 1.)):
                 y = torch.zeros_like(labels)
                 y[mask] = labels[mask]
             
-            last = (1 - self.alpha) * y
```
Comment (mufeili) [inline examples/pytorch/label_propagation/model.py:44]: Is it possible to replace `LabelPropagation` in this example with the built-in `LabelPropagation` module?

File: python/dgl/nn/pytorch/utils.py [hunk 2]
```diff
@@ -395,3 +397,113 @@ def forward(self, feat_list):
...
             return (stacked_feat_list * alpha.unsqueeze(-1)).sum(dim=1)
+
+class LabelPropagation(nn.Module):
+    r"""Label Propagation from `Learning from Labeled and Unlabeled Data with Label
+    Propagation <http://mlg.eng.cam.ac.uk/zoubin/papers/CMU-CALD-02-107.pdf>`__
+
+    .. math::
+    \mathbf{Y}^{(t+1)} = \alpha \cdot \tilde{A} \mathbf{Y}^{(t)} + (1 - \alpha)
+    \mathbf{Y}^{(0)},
+    where unlabeled data is inferred by labeled data via propagation. :math:`\alpha`
```
Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:408]: Add an empty line between L407 and L408 to ensure proper rendering.

Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:408]: "where unlabeled data is inferred by labeled data via propagation" -> "where unlabeled data is initially set to zero and inferred from labeled data via propagation"

File: python/dgl/nn/pytorch/utils.py [hunk 3]
```diff
@@ -395,3 +397,113 @@ def forward(self, feat_list):
...
+    is a weight parameter for balancing between updated labels and initial labels.
+    :math:`\tilde{A}` denotes the normalized adjancency matrix.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:410]: "adjancency" -> "adjacency"

File: python/dgl/nn/pytorch/utils.py [hunk 4]
```diff
@@ -395,3 +397,113 @@ def forward(self, feat_list):
...
+        The number of propagation steps.
+    alpha : float
+        The :math:`\alpha` coefficient in range [0, 1].
+    norm_type : str, optional
+        The type of normalization applied to the adjacency matrix, must be one of the
+        following choices:
+
+        * ``row``: row-normalized adjacency as :math:`D^{-1}A`
+
+        * ``sym``: symmetric normalized adjacency as :math:`D^{-1/2}AD^{-1/2}`
```
Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:424]: "symmetric" -> "symmetrically"

File: python/dgl/nn/pytorch/utils.py [hunk 5]
```diff
@@ -395,3 +397,113 @@ def forward(self, feat_list):
...
+
+        Default: 'sym'.
+    multi_label : bool, optional
```
Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:427]: My bad. I'm now thinking if we should have this flag as it harms flexibility. Probably it's still better to have two separate arguments `clamp` and `normalize`. What do you think? @rudongyu @jermainewang

**Comment by mufeili** [comment]
I remember `DGN` requires an optional argument for random flipping in computing Laplacian positional encodings. Could you also extend `LapPE`? @rudongyu

File: python/dgl/nn/pytorch/utils.py [hunk 6]
```diff
@@ -395,3 +397,113 @@ def forward(self, feat_list):
...
+        A bool flag to indicate the classification task is multi-label or multi-class.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:428]: "indicate the" -> "indicate whether the"

File: python/dgl/nn/pytorch/utils.py [hunk 7]
```diff
@@ -395,3 +397,113 @@ def forward(self, feat_list):
...
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The input graph.
+        labels : torch.Tensor
+            The input labels of shape :math:`(N, 1)` or :math:`(N,)` with dtype as
+            torch.long; or input labels of shape :math:`(N, C)` with dtype as
+            torch.float. :math:`N` and :math:`C` are the number of nodes and the number
+            of categories, respectively.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:466]: How about multitask classification?

File: python/dgl/nn/pytorch/utils.py [hunk 8]
```diff
@@ -395,3 +397,113 @@ def forward(self, feat_list):
...
+        mask : torch.Tensor
+            The bool indicators of shape :math:`(N,)` with True denoting labeled nodes.
+            Default: None, indicating all nodes are labeled.
+
+        Returns
+        -------
+        The propagated node label tensor of shape :math:`(N, C)` with float type.
+        """
+        with g.local_scope():
```
Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:475]: It seems that the scenario of multitask classification is not handled here.

File: python/dgl/nn/pytorch/utils.py [hunk 9]
```diff
@@ -395,3 +397,113 @@ def forward(self, feat_list):
...
+
+            y = labels
+            if mask is not None:
+                y = th.zeros_like(labels)
+                y[mask] = labels[mask]
+
+            init = (1 - self.alpha) * y
+            degs = g.in_degrees().float().clamp(min=1)
+            if self.norm_type == 'sym':
+                norm = th.pow(degs, -0.5).to(labels.device).unsqueeze(1)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/utils.py:487]: What if the graph is directed, i.e., the existence of (i, j) does not imply that of (j, i).

File: tests/pytorch/test_nn.py [hunk 10]
```diff
@@ -1483,3 +1483,42 @@ def test_pna_conv(in_size, out_size, aggregators, scalers, delta,
...
+@pytest.mark.parametrize('reset', [True, False])
+def test_label_prop(k, alpha, norm_type, multi_label, reset):
+    dev = F.ctx()
+    num_nodes = 5
+    num_edges = 20
+    g = dgl.rand_graph(num_nodes, num_edges).to(dev)
+    labels = th.tensor([0, 2, 1, 3, 0]).long().to(dev)
+    mask = th.tensor([0, 1, 1, 1, 0]).bool().to(dev)
+    model = nn.LabelPropagation(k, alpha, norm_type, multi_label, reset)
+    model(g, labels, mask)
```
Comment (mufeili) [inline tests/pytorch/test_nn.py:1500]: Need test for two cases:
1) multiclass classification + one-hot labels
2) multitask classification

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+                h.append(aggregator(msg))
+        h = torch.cat(h, dim=1)
+        h = torch.cat([
+            scaler(h, D=degree, delta=self.delta) if scaler is not scale_identity else h
+            for scaler in self.scalers
+        ], dim=1)
+        return {'h_neigh': h}
+
+class DGNConv(PNAConv):
+    r"""Directional Graph Network Layer from `Directional Graph Networks
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:68]: You do not need to mention PNA in the doc. The goal is to make the doc self-contained. If users need to be aware of the connection between PNA and DGN, they will do so regardless of the doc.

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    :math:`M` and :math:`U` are MLPs, taking the concatenation of input for computing
+    output features. :math:`\oplus` represents the combination of various aggregators
+    and scalers. Aggregators aggregate messages from neighbours and scalers scale the
+    aggregated messages in different ways. :math:`\oplus` concatenates the output features
+    of each combination.
+
+    The output of multiple towers are concatenated and fed into a linear mixing layer for the
+    final output.
+    
+    Two additional directional aggregators are supported in DGN according to the vector field
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:88]: You can simply say that "DGN introduces two special aggregators." to avoid mentioning PNA.

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    :math:`F`, which is defined as the gradient of the low-frequency eigenvectors of graph laplacian.
+    
+    The directional average aggregator is defined as
+    :math:`h_i' = \sum_{j\in\mathcal{N}(i)}\frac{|F_{i,j}|\cdot h_j}{||F_{i,:}||_1+\epsilon}`
+    
+    The directional derivative aggregator is defined as
+    :math:`h_i' = \sum_{j\in\mathcal{N}(i)}\frac{F_{i,j}\cdot h_j}{||F_{i,:}||_1+\epsilon}
+    -h_i\cdot\sum_{j\in\mathcal{N}(i)}\frac{F_{i,j}}{||F_{i,:}||_1+\epsilon}`
+
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:97]: Explain what is \epsilon here.

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    Directional Graph Networks allow a better representation of the anisotropic features in
+    different physical or biological problems
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:99]: L98-99 will be moved to the release note instead.

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+        :math:`(E[(X-E[X])^n])^{1/n}`
+        
+        * ``dir{k}-av``: directional average aggregation with directions defined by the k-th
+        smallest eigenvectors. k can be selected from 1, 2, 3.
+        
+        * ``dir{k}-dx``: directional derivative aggregation with directions defined by the k-th
+        smallest eigenvectors. k can be selected from 1, 2, 3.
+
+        Note that using directional aggregation requires the LaplacianPE transform on the input
+        graph for eigenvectors computation (the PE size must be >= k above).
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:133]: "eigenvectors computation" -> "eigenvector computation"

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+                self.tower_in_size, self.tower_out_size,
+                aggregators, scalers, delta,
+                dropout=dropout, edge_feat_size=edge_feat_size
+            ) for _ in range(num_towers)
+        ])
+
+        self.mixing_layer = nn.Sequential(
+            nn.Linear(out_size, out_size),
+            nn.LeakyReLU()
+        )
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:200]: You need to define a forward function that directly takes computed Laplacian eigenvectors.

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+        * ``identity``: no scaling
+
+        * ``amplification``: multiply the aggregated message by :math:`\log(d+1)/\delta`,
+        where :math:`d` is the in-degree of the node.
+
+        * ``attenuation``: multiply the aggregated message by :math:`\delta/\log(d+1)`
+    delta: float
+        The degree-related normalization factor computed over the training set, used by scalers
+        for normalization. :math:`E[\log(d+1)]`, where :math:`d` is the degree for each node
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:145]: To be consistent with the description of scalers, "degree" -> "in-degree"?

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+    >>> import torch as th
+    >>> from dgl.nn import DGNConv
+    >>> from dgl import LaplacianPE
+    >>>
+    >>> transform = LaplacianPE(k=3)
+    >>> g = dgl.graph(([0,1,2,3,2,5], [1,2,3,4,0,3]))
+    >>> g = transform(g)
+    >>> feat = th.ones(6, 10)
+    >>> conv = DGNConv(10, 10, ['dir1-av', 'dir1-dx', 'sum'], ['identity', 'amplification'], 2.5)
+    >>> ret = conv(g, feat)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:171]: Need to pass the Laplacian eigenvectors in `conv(...)`.

File: python/dgl/nn/pytorch/conv/dgnconv.py [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+        degree = msg.size(1)
+
+        h = []
+        for aggregator in self.aggregators:
+            agg_name = aggregator.__name__
+            if agg_name.startswith('dir'):
+                if agg_name.endswith('av'):
+                    h.append(aggregator(msg, eig_s, eig_d))
+                else:
+                    h.append(aggregator(msg, eig_s, eig_d, h_in))
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/dgnconv.py:57]: It seems that `eig_idx` is not passed.

**Review by mufeili — CHANGES_REQUESTED** [review]
Done a first pass

File: python/dgl/nn/pytorch/utils.py (hunk 5 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/utils.py]: Sure, i think so. Changed.

File: python/dgl/nn/pytorch/utils.py (hunk 9 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/utils.py]: Adapted to both directed and undirected graphs. Need further check.

File: python/dgl/nn/pytorch/conv/dgnconv.py (hunk 16 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/conv/dgnconv.py]: That may affect the inheritance from PNA a lot, requiring overloaded forward functions for both DGN and DGNTower. I think it is unnecessary. I prefer the current way of carrying eigenvectors in the graph with a key "eig". How do you think of it?

Comment (rudongyu) [reply python/dgl/nn/pytorch/conv/dgnconv.py]: That may affect the inheritance from PNA a lot, requiring overloaded forward functions for both DGN and DGNTower. I think it is unnecessary. I prefer the current way of carrying eigenvectors in the graph with a key "eig". How do you think of it?

File: python/dgl/nn/pytorch/conv/dgnconv.py (hunk 18 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/conv/dgnconv.py:154]: ditto.

**Review by mufeili — APPROVED** [review]
